db/migrate/20231114223004_add_my_settings_view_to_roles.rb

```diff
@@ -0,0 +1,48 @@
+class AddMySettingsViewToRoles < ActiveRecord::Migration[6.1]
+  class MiqUserRole < ActiveRecord::Base
+    has_and_belongs_to_many :miq_product_features, :join_table => :miq_roles_features, :class_name => "AddMySettingsViewToRoles::MiqProductFeature"
+  end
+
+  class MiqProductFeature < ActiveRecord::Base; end
+
+  def up
+    return unless MiqUserRole.exists?(:read_only => false)
+
+    say_with_time("Adding my_settings_view to custom user roles") do
+      my_settings_features = MiqProductFeature.where("identifier LIKE ?", "my_settings_%").where.not(:identifier => "my_settings_view")
+      my_settings_view = nil
+
+      affected_user_roles(my_settings_features).each do |user_role|
+        product_features = user_role.miq_product_features.map(&:identifier)
+
+        next if product_features.include?("my_settings") || product_features.include?("my_settings_view")
```

Comment:
```suggestion
      my_settings_id = MiqProductFeature.find_by(:identifier => "my_settings").id

      my_settings_features = MiqProductFeature.where("identifier LIKE ?", "my_settings_%").select(:identifier)
      my_settings_features_excludes = 
      my_settings_view = nil

      affected_user_roles(my_settings_features)
        .where.not(:id => affected_user_roles(%w[my_settings my_settings_view], read_only: true))
        .each do |user_role|
```

`affected_user_roles` says all roles that have features of the given identity.

So we want to bring back all roles that have certain features
but not the roles that have certain features (the guard clause)

Reply:
Interesting - let me try

Reply:
Ok, updated.  The `read_only: true` was wrong in the example there, so you won't see that included, but the general direction of a subquerey where not worked.